### ISSUE TITLE
reading from stream fails for stdin with ENOTCONN

### ISSUE THREAD

**Issue by joerg-krause** [luvit/luvit#1076]
Since libuv commit https://github.com/libuv/libuv/commit/8f9ba2a59709a36502fff13ffcb6a36b04bd8506 reading from a stream not flagged as readable fails with [ENOTCONN](https://github.com/libuv/libuv/blob/d16e6094e1eb3b0b5981ef1dd7e03ec4d466944d/src/unix/stream.c#L1559-L1560).

EDIT: Add a more detailed explanation:

This has the consequence, that the `stdin` stream will not properly get closed, so the [exitHandler()](https://github.com/luvit/luvit/blob/001cb628e4aca5c034f6834613989e3bd9925104/deps/childprocess.lua#L211-L234) in childprocess.lua is called not often enough (it is only called from the [data handler of the stdout stream](https://github.com/luvit/luvit/blob/001cb628e4aca5c034f6834613989e3bd9925104/deps/childprocess.lua#L252-L259), but not from the [close handler](exitHandler)) and therefore the `close` signal and the callback function of `exec()` is never executed.

**Comment by joerg-krause** [comment]
From my limited understanding I would like to discuss two options I see:

1. Check the return value of `uv.read_start()`.
2. Drop `stdin` as `stdio` in  `spawn()`.

The first option means:

```diff
diff --git a/deps/net.lua b/deps/net.lua
index f5ffe82..ded3b92 100644
--- a/deps/net.lua
+++ b/deps/net.lua
@@ -130,7 +130,10 @@ function Socket:_read(n)
     self:once('connect', utils.bind(self._read, self, n))
   elseif not self._reading then
     self._reading = true
-    uv.read_start(self._handle, onRead)
+    local _, err = uv.read_start(self._handle, onRead)
+    if err ~= nil then
+      return self:destroy(err)
+    end
   end
 end
```
In this case, for as for `stdin` libuv return `ENOTCONN`, `stdin` is closed.

The second options means:
```diff
diff --git a/deps/childprocess.lua b/deps/childprocess.lua
index 1361675..d4d5d26 100644
--- a/deps/childprocess.lua
+++ b/deps/childprocess.lua
@@ -112,11 +112,10 @@ local function spawn(command, args, options)
     if stdio[3] then options.stdio[3]:once('close', maybeClose) end
     closesGot = countStdio(stdio)
   else
-    stdin = net.Socket:new({ handle = uv.new_pipe(false) })
+    stdin = nil
     stdout = net.Socket:new({ handle = uv.new_pipe(false) })
     stderr = net.Socket:new({ handle = uv.new_pipe(false) })
-    stdio = { stdin._handle, stdout._handle, stderr._handle}
-    stdin:once('close', maybeClose)
+    stdio = { nil, stdout._handle, stderr._handle}
     stdout:once('close', maybeClose)
     stderr:once('close', maybeClose)
     closesGot = countStdio(stdio)
```

Note, that for the second option the test case for `test-process-*` needs to be fixed as `stdin` might be `nil`.

The first option is actually error handling, the second option is error avoidance.

What do you thing?

**Comment by joerg-krause** [comment]
Fixed by 9d14161ee72a7d85aa1c8bdb3435427514738786.
